lib/vrl/compiler/src/type_def.rs

```diff
@@ -926,6 +952,197 @@ impl TypeDef {
         self
     }
 
+    fn update_segment<'a, I>(
+        kind: &KindInfo,
+        mut path: std::iter::Peekable<I>,
+        newkind: &KindInfo,
+    ) -> KindInfo
+    where
+        I: std::iter::Iterator<Item = &'a SegmentBuf> + Clone,
+    {
+        match kind {
+            KindInfo::Unknown => {
+                // The kind is unknown and thus there is nothing to remove.
+                kind.clone()
+            }
+            KindInfo::Known(kinds) => KindInfo::Known(
+                kinds
+                    .iter()
+                    .map(|kind| match (kind, path.next(), path.peek()) {
+                        (
```

Comment:
Given the large match logic here, it might make sense to add some comments to each match arm to explain when it matches and what it's supposed to do if it matches.

Reply:
I refactored a bit so the match logic is simpler, so hopefully it negates the need for comments on each arm.